attestation-agent/attestation-agent/src/eventlog/mod.rs

```diff
@@ -62,124 +93,141 @@ impl EventLog {
-        };
-        eventlog
-            .extend_init_entry()
-            .await
-            .context("extend INIT entry")?;
-        Ok(eventlog)
+        })
+        }
     }
 
-    pub async fn extend_entry(&mut self, log_entry: LogEntry<'_>, pcr: u64) -> Result<()> {
-        let digest = log_entry.digest_with(self.alg);
-        // The order must be ensured to keep consistency. s.t. first write AAEL
-        // and then extend RTMR.
-        self.writer.append(&log_entry).context("write log entry")?;
-        self.rtmr_extender
-            .extend_runtime_measurement(digest, pcr)
-            .await?;
+    /// Record the event and the target digest into cache file before write, this would do
+    /// help when there is a crush between extending PCR and logging event.
+    fn wal_cache(&self, wal_cache: WalCache) -> Result<()> {
+        let mut file = File::create(WAL_CACHE)?;
+        file.write_all(&wal_cache.event_offset.to_be_bytes())?;
+        file.write_all(wal_cache.expected_pcr.as_ref())?;
+        file.write_all(wal_cache.event_data.as_ref())?;
+        file.sync_data()?;
+        Ok(())
+    }
 
+    /// Remove the wal cache file.
+    pub fn clean_wal_cache(&self) -> Result<()> {
+        remove_file(WAL_CACHE)?;
         Ok(())
     }
 
-    pub async fn extend_init_entry(&mut self) -> Result<()> {
-        let pcr = self.rtmr_extender.get_runtime_measurement(self.pcr).await?;
-        let init_value = hex::encode(pcr);
-        let init_value = format!("{:0>width$}", init_value, width = self.alg.digest_len());
-        let init_entry = LogEntry::Init {
-            hash_alg: self.alg,
-            value: &init_value,
+    /// Try to read the wal cache file.
+    fn read_wal_cache(digest_len: usize) -> Result<Option<WalCache>> {
+        if !Path::new(WAL_CACHE).exists() {
+            return Ok(None);
+        }
+        let mut file = File::open(WAL_CACHE)?;
+        let mut event_offset = [0u8; 8];
+        file.read_exact(&mut event_offset)?;
+        let event_offset = u64::from_le_bytes(event_offset);
+
+        let mut expected_pcr = vec![0u8; digest_len];
+        file.read_exact(&mut expected_pcr)?;
+
+        let mut event_data = String::new();
+        let _ = file.read_to_string(&mut event_data)?;
+
+        Ok(Some(WalCache {
+            expected_pcr,
+            event_data,
+            event_offset,
+        }))
+    }
+
+    /// This an atomic operation, which will both extend the PCR and write the eventlog to disk.
+    /// We should bring in a little transection mechanism (WAL, Write Ahead Log) to make sure the atomicity.
+    ///
+    /// The logical order:
+    ///
+    /// 1. write expected PCR and AAEL event data to WAL cache file on disk
+    /// 2. Extend PCR
+    /// 3. Write eventlog entry to AAEL file on disk
+    /// 4. delete WAL cache file on disk
+    ///
+    /// Once kernel provides a more robust way (atomic) to maintain an eventlog,
+    /// we can remove the WAL cache file mechanism.
```

Comment:
FYI, the upstream tsm RTMR implementation provides a tracepoint and userspace can hook into to trace all the measurements. it has no ABI to store the log content tho.